### PR TITLE
[SPARK-35560][SQL] Remove redundant subexpression evaluation in nested subexpressions

### PR DESCRIPTION
**Pull request by viirya**
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error message, please read the guideline first:
     https://spark.apache.org/error-message-guidelines.html
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This patch proposes to improve subexpression evaluation under whole-stage codegen for the cases of nested subexpressions.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

In the cases of nested subexpressions, whole-stage codegen's subexpression elimination will do redundant subexpression evaluation. We should reduce it. For example, if we have two sub-exprs:

1. `simpleUDF($"id")`
2. `functions.length(simpleUDF($"id"))`

We should only evaluate `simpleUDF($"id")` once, i.e.

```java
subExpr1 = simpleUDF($"id");
subExpr2 = functions.length(subExpr1);
```

Snippets of generated codes:

Before:
```java
/* 040 */   private int project_subExpr_1(long project_expr_0_0) {                                                                                                           
/* 041 */     boolean project_isNull_6 = false;                                                                                                                              
/* 042 */     UTF8String project_value_6 = null;                                      
/* 043 */     if (!false) {                                                                                                                                                  
/* 044 */       project_value_6 = UTF8String.fromString(String.valueOf(project_expr_0_0));                                                                                   
/* 045 */     }                                                                                                                                                              
/* 046 */                                                                                                                                                                    
/* 047 */     Object project_arg_1 = null;                                            
/* 048 */     if (project_isNull_6) {                                                                                                                                        
/* 049 */       project_arg_1 = ((scala.Function1[]) references[3] /* converters */)[0].apply(null);                                                                         
/* 050 */     } else {                                                                                                                                                       
/* 051 */       project_arg_1 = ((scala.Function1[]) references[3] /* converters */)[0].apply(project_value_6);                                                              /* 052 */     }                                                                                                                                                              
/* 053 */                                                                                                                                                                    
/* 054 */     UTF8String project_result_1 = null;                                                                                                                            /* 055 */     try {                                                                                                                                                          /* 056 */       project_result_1 = (UTF8String)((scala.Function1[]) references[3] /* converters */)[1].apply(((scala.Function1) references[4] /* udf */).apply(project_arg_1)
);                                                                                                                                                                           
/* 057 */     } catch (Throwable e) {                                                 
/* 058 */       throw QueryExecutionErrors.failedExecuteUserDefinedFunctionError(                                                                                            
/* 059 */         "DataFrameSuite$$Lambda$6418/1507986601", "string", "string", e);                                                                                          
/* 060 */     }                                                                       
/* 061 */                                                                             
/* 062 */     boolean project_isNull_5 = project_result_1 == null;
/* 063 */     UTF8String project_value_5 = null;
/* 064 */     if (!project_isNull_5) {
/* 065 */       project_value_5 = project_result_1;
/* 066 */     }                                                                                                                                                              
/* 067 */     boolean project_isNull_4 = project_isNull_5;                                                                                                                   
/* 068 */     int project_value_4 = -1;                                                                                                                                      
/* 069 */                                                                                                                                                                    
/* 070 */     if (!project_isNull_5) {                                                                                                                                       
/* 071 */       project_value_4 = (project_value_5).numChars();                                                                                                              
/* 072 */     }                                                                                                                                                              
/* 073 */     project_subExprIsNull_1 = project_isNull_4;                                                                                                                    
/* 074 */     return project_value_4;                                                                                                                                        
/* 075 */   }    
...                          
/* 149 */   private UTF8String project_subExpr_0(long project_expr_0_0) {                                                                                                    
/* 150 */     boolean project_isNull_2 = false;                                                                                                                              
/* 151 */     UTF8String project_value_2 = null;                                                                                                                             
/* 152 */     if (!false) {                                                                                                                                                  
/* 153 */       project_value_2 = UTF8String.fromString(String.valueOf(project_expr_0_0));                                                                                   
/* 154 */     }                                                                                                                                                              
/* 155 */                                                                                                                                                                    
/* 156 */     Object project_arg_0 = null;                                                                                                                                   
/* 157 */     if (project_isNull_2) {                                                                                                                                        
/* 158 */       project_arg_0 = ((scala.Function1[]) references[1] /* converters */)[0].apply(null);                                                                         
/* 159 */     } else {                                                                                                                                                       
/* 160 */       project_arg_0 = ((scala.Function1[]) references[1] /* converters */)[0].apply(project_value_2);                                                              
/* 161 */     }                                                                       
/* 162 */                                                                                                                                                                    
/* 163 */     UTF8String project_result_0 = null;                                                                                                                            
/* 164 */     try {                                                                                                                                                          
/* 165 */       project_result_0 = (UTF8String)((scala.Function1[]) references[1] /* converters */)[1].apply(((scala.Function1) references[2] /* udf */).apply(project_arg_0)
);                                                                                    
/* 166 */     } catch (Throwable e) {                                                                                                                                        
/* 167 */       throw QueryExecutionErrors.failedExecuteUserDefinedFunctionError(                                                                                            
/* 168 */         "DataFrameSuite$$Lambda$6418/1507986601", "string", "string", e);                                                                                          
/* 169 */     }                                                                                                                                                              
/* 170 */                                                                                                                                                                    
/* 171 */     boolean project_isNull_1 = project_result_0 == null;                                                                                                           /* 172 */     UTF8String project_value_1 = null;                                                                                                                             /* 173 */     if (!project_isNull_1) {                                                                                                                                       /* 174 */       project_value_1 = project_result_0;                                                                                                                          
/* 175 */     }                                                                                                                                                              
/* 176 */     project_subExprIsNull_0 = project_isNull_1;                             
/* 177 */     return project_value_1;                                                                                                                                        
/* 178 */   }                                     
```

After:
```java
/* 041 */   private void project_subExpr_1(long project_expr_0_0) {                                                                                                          
/* 042 */     boolean project_isNull_8 = project_subExprIsNull_0;                                                                                                            
/* 043 */     int project_value_8 = -1;                                                                                                                                      
/* 044 */                                                                                                                                                                    
/* 045 */     if (!project_subExprIsNull_0) {                                                                                                                                
/* 046 */       project_value_8 = (project_mutableStateArray_0[0]).numChars();                                                                                               
/* 047 */     }                                                                                                                                                              
/* 048 */     project_subExprIsNull_1 = project_isNull_8;                                                                                                                    
/* 049 */     project_subExprValue_0 = project_value_8;                                                                                                                      
/* 050 */   }                                                                                                                                           
/* 056 */                                                                                                                                           
...                                                                                                                                                                                                                                                                                                                    
/* 123 */                                                                                                                                                                    
/* 124 */   private void project_subExpr_0(long project_expr_0_0) {                                                                                                          
/* 125 */     boolean project_isNull_6 = false;                                                                                                                              
/* 126 */     UTF8String project_value_6 = null;                                                                                                                             
/* 127 */     if (!false) {                                                                                                                                                  
/* 128 */       project_value_6 = UTF8String.fromString(String.valueOf(project_expr_0_0));                                                                                   
/* 129 */     }                                                                                                                                                              
/* 130 */                                                                             
/* 131 */     Object project_arg_1 = null;                                            
/* 132 */     if (project_isNull_6) {                                                 
/* 133 */       project_arg_1 = ((scala.Function1[]) references[3] /* converters */)[0].apply(null);                                                                         
/* 134 */     } else {                                                                
/* 135 */       project_arg_1 = ((scala.Function1[]) references[3] /* converters */)[0].apply(project_value_6);
/* 136 */     }                                                                       
/* 137 */                                                                                                                                                                    
/* 138 */     UTF8String project_result_1 = null;                                                                                                                            
/* 139 */     try {                                                                   
/* 140 */       project_result_1 = (UTF8String)((scala.Function1[]) references[3] /* converters */)[1].apply(((scala.Function1) references[4] /* udf */).apply(project_arg_1)
);                                                                                    
/* 141 */     } catch (Throwable e) {                                                 
/* 142 */       throw QueryExecutionErrors.failedExecuteUserDefinedFunctionError(     
/* 143 */         "DataFrameSuite$$Lambda$6430/2004847941", "string", "string", e);
/* 144 */     }                                                                                                                                                              
/* 145 */                                                                             
/* 146 */     boolean project_isNull_5 = project_result_1 == null;
/* 147 */     UTF8String project_value_5 = null;                                      
/* 148 */     if (!project_isNull_5) {                                                                                                                                       
/* 149 */       project_value_5 = project_result_1;                                 
/* 150 */     }                                                                       
/* 151 */     project_subExprIsNull_0 = project_isNull_5;          
/* 152 */     project_mutableStateArray_0[0] = project_value_5;                                                                                                              
/* 153 */   }                                                          
```

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Unit test.

### CODE REVIEW

**Comment by Kimahriman** [comment]
Two questions just to see if I understand things correctly: this builds upon/relies upon the common expressions being sorted from your previous MR right? And theoretically if https://github.com/apache/spark/pull/32559 hadn't been fixed, this would sort of address that issue by creating a subexpression function for each level that builds on the previous instead of starting from scratch?

**Comment by viirya** [comment]
> this builds upon/relies upon the common expressions being sorted from your previous MR right?

Yes, if you're asking if this is built on previous PR that sorts subexpressions.

> And theoretically if #32559 hadn't been fixed, this would sort of address that issue by creating a subexpression function for each level that builds on the previous instead of starting from scratch?

Yea, if #32559 is not fixed, this could somehow relieve that but it still suffers from redundant subexpression call if it is split case.

**Comment by viirya** [comment]
cc @maropu @cloud-fan

**Comment by Kimahriman** [comment]
> Yes, if you're asking if this is built on previous PR that sorts subexpressions.

Hah whoops I spend most of my day dealing with merge requests not pull requests 😅

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -1039,21 +1039,25 @@ class CodegenContext extends Logging {
...
-    lazy val commonExprVals = commonExprs.map(_.head.genCode(this))
 
-    lazy val nonSplitExprCode = {
-      commonExprs.zip(commonExprVals).map { case (exprs, eval) =>
-        // Generate the code for this expression tree.
-        val state = SubExprEliminationState(eval.isNull, eval.value)
-        exprs.foreach(localSubExprEliminationExprs.put(_, state))
+    val nonSplitExprCode = {
+      commonExprs.map { exprs =>
+        val eval = withSubExprEliminationExprs(localSubExprEliminationExprsForNonSplit.toMap) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1054]: what does this recursive call of `withSubExprEliminationExprs` give us?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -1068,11 +1072,19 @@ class CodegenContext extends Logging {
...
+          mutable.HashMap.empty[Expression, SubExprEliminationState]
+
+        val splitCodes = commonExprs.zipWithIndex.map { case (exprs, i) =>
           val expr = exprs.head
-          val eval = commonExprVals(i)
+          val eval = withSubExprEliminationExprs(localSubExprEliminationExprs.toMap) {
+            Seq(expr.genCode(this))
+          }.head
+
+          val value = addMutableState(javaType(expr.dataType), "subExprValue")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1087]: why does this have to be a mutable state now?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1087]: Use the example in the description to explain. For the two subexpressions:

1. `simpleUDF($"id")`
2. `functions.length(simpleUDF($"id"))`


Previously we evaluate them independently, i.e.,

```
String subExpr1 = simpleUDF($"id");
Int subExpr2 = functions.length(simpleUDF($"id"));
```

Now we remove redundant evaluation of nested subexpressions:
```
String subExpr1 = simpleUDF($"id");
Int subExpr2 = functions.length(subExpr1);
```

If we need to split the functions, when we evaluate `functions.length`, it needs access of `subExpr1`. We have two choices. One is to add `subExpr1` to the function parameter list of the split function for `functions.length`. Another one is to use mutable state.

To add it to parameter list will complicate the way we compute parameter length. That's said we need to link nested subexpression relations and get the correct parameters. Seems to me it is not worth doing that.

Currently I choose the simpler approach that is to use mutable state.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1054]: For each set of common expressions, `withSubExprEliminationExprs` only called once so I think it is not actually a recursive call?

`withSubExprEliminationExprs` takes the given map used for subexpression elimination to replace common expression during expression codegen in the closure. It returns evaluated expression code (value/isNull/code).

For the two subexpressions as example:

1. `simpleUDF($"id")`
2. `functions.length(simpleUDF($"id"))`

1st round `withSubExprEliminationExprs`:

The map is empty.
Gen code for `simpleUDF($"id")`.
Put it into the map => (`simpleUDF($"id")` -> gen-ed code)

2nd round `withSubExprEliminationExprs`:

Gen code for `functions.length(simpleUDF($"id"))`.
Looking at the map and replace common expression `simpleUDF($"id")` as gen-ed code.
Put it into the map => (`simpleUDF($"id")` -> gen-ed code, `functions.length(simpleUDF($"id"))` -> gen-ed code)

The map will be used later for subexpression elimination.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 3]
```diff
@@ -2882,6 +2882,31 @@ class DataFrameSuite extends QueryTest
...
+      withSQLConf(SQLConf.CODEGEN_METHOD_SPLIT_THRESHOLD.key -> splitThreshold.toString) {
+        val accum = sparkContext.longAccumulator("call")
+        val simpleUDF = udf((s: String) => {
+          accum.add(1)
+          s
+        })
+
+        // Common exprs:
+        //  1. simpleUDF($"id")
+        //  2. functions.length(simpleUDF($"id"))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2897]: Q: What if a tree has more deeply-nested common exprs? The current logic can work well? e.g., I thought it like this;
```
        // subExpr1 = simpleUDF($"id");
        // subExpr2 = functions.length(subExpr1);
        // subExpr3 = functions.xxxx(subExpr2);
        // subExpr4 = ...
```

**Comment by maropu** [comment]
I left one question and the other part looks fine.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2897]: Yes, this is actually the cases this logic to deal with. Previous common expression gen-ed codes will put into the map. The code generator looks up into the map when generating code for later common expressions to replace the semantic-equal expression with gen-ed code value.

**Comment by viirya** [comment]
Thanks @cloud-fan @maropu. I will merge this later if no more comments.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 4]
```diff
@@ -2882,6 +2882,31 @@ class DataFrameSuite extends QueryTest
     df2.collect()
     assert(accum.value == 15)
   }
+
+  test("SPARK-35560: Remove redundant subexpression evaluation in nested subexpressions") {
+    Seq(1, Int.MaxValue).foreach { splitThreshold =>
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2887]: How long does it take to run this test case?

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by viirya** [comment]
Thank you @dongjoon-hyun

**Comment by viirya** [comment]
Thanks all! Merging to master.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 3 above)
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2897]: Nice

**Pull request closed without merging**